File: src/utils/glob.ts
import {glob as nodeGlob} from "node:fs/promises";
import path, {matchesGlob} from "node:path";
import {toPosix} from "./path.js";

export interface GlobOptions {
    cwd: string,
    ignore?: string[],
}

export async function glob(patterns: string[], options: GlobOptions) {
    const fileNames = await Array.fromAsync(
        nodeGlob(patterns, {
            cwd: options.cwd,
            withFileTypes: true,
            exclude: it => {
                const fileName = path.resolve(it.parentPath, it.name)

                return (options.ignore ?? [])
                    .some(it => matchesGlob(fileName, it))
            }
        })
    )

    return fileNames.map(it => toPosix(path.resolve(it.parentPath, it.name)))
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/util/glob.kt
package io.github.sgrishchenko.karakum.util

import js.array.JsArrays
import js.promise.await
import node.fs.Dirent
import node.fs.GlobOptionsWithFileTypes
import node.path.path

suspend fun glob(
    patterns: List<String>,
    cwd: String,
    ignore: List<String> = emptyList()
): List<String> {
    val fileNames = JsArrays.fromAsync(
        node.fs.glob(patterns.toTypedArray(), GlobOptionsWithFileTypes(
            cwd = cwd,
            requiredWithFileTypes = true,
            exclude = { file: Dirent<String> ->
                val fileName = path.resolve(file.parentPath, file.name)

                ignore.any { path.matchesGlob(fileName, it) }
            },
        ))
    ).await()

    return fileNames
        .map { it.unsafeCast<Dirent<String>>() }
        .map { toPosix(path.resolve(it.parentPath, it.name)) }
}
